File: NeatUpload/ProgressBar.js
/*
NeatUpload - an HttpModule and User Controls for uploading large files
Copyright (C) 2005  Dean Brettle

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
if (!Array.prototype.push)
{
	Array.prototype.push = function() {
		for (var i = 0; i < arguments.length; i++)
			this[this.length] = arguments[i];
		return this.length;
	};
}

if (!Array.prototype.unshift)
{
	Array.prototype.unshift = function() {
		this.reverse();
		for (var i = 0; i < arguments.length; i++)
			this[this.length] = arguments[i];
		this.reverse();
		return this.length;
	};
}

if (!Function.prototype.call)
{
	Function.prototype.call = function() {
		var obj = arguments[0];
		obj._NeatUpload_tmpFunc = this;
		var argList = '';
		for (var i = 1; i < arguments.length; i++)
		{
			argList += 'arguments[' + i + ']';
			if (i < arguments.length - 1)
				argList += ',';
		}
		var result = eval('obj._NeatUpload_tmpFunc(' + argList + ')');
		obj._NeatUpload_tmpFunc = null;
		return result;
	};
}

function NeatUploadPB(id, popupDisplayStatement, inline, displayFunc, triggerIDs, autoStartCondition)
{
	if (!document.getElementById)
		return;
	var pb = this;

	this.OnUnloadHandlers = new Array();
	var origOnUnload = window.onunload;
	this.OnUnloadHandlers.push(function () { window.onunload = origOnUnload; });
	window.onunload = this.CombineHandlers(window.onunload, function() { return pb.OnUnload(); });

	var elem = document.getElementById(id);
	if (!elem)
		elem = document.getElementById(id + '_NeatUpload_dummyspan');
	var formElem = elem;
	while (formElem && formElem.tagName.toLowerCase() != "form")
	{
		formElem = formElem.parentNode;
	}
	this.OnSubmitForm = function()
	{
		return formElem.NeatUpload_OnSubmit();
	}
	var origOnSubmit = formElem.onsubmit;
	this.OnUnloadHandlers.push(function () { formElem.onsubmit = origOnSubmit; });
	formElem.onsubmit = this.CombineHandlers(formElem.onsubmit, this.OnSubmitForm);
	
	var fallbackLink = document.getElementById(id + '_fallback_link');
	if (fallbackLink)
		fallbackLink.setAttribute('href', 'javascript:' + popupDisplayStatement);
	this.TriggerIDs = new Object();
	this.TriggerIDs.length = 0;
	this.Display = displayFunc;
	this.FormElem = formElem;
	this.AutoStartCondition = autoStartCondition;
	this.AddSubmitHandler(pb.FormElem, !inline, function () {
		// If trigger controls were specified for this progress bar and the trigger is not 
		// specified for *any* progress bar, then clear the filenames.
		if (NeatUpload_LastEventSource
		    && pb.TriggerIDs.length
		    && !pb.IsElemWithin(NeatUpload_LastEventSource, NeatUpload_TriggerIDs))
		{
			return pb.ClearFileInputs(pb.FormElem);
		}
		// If there are files to upload and either no trigger controls were specified for this progress bar or
		// a specified trigger control was triggered, then start the progress display.
		if (pb.EvaluateAutoStartCondition()
			&& (!pb.TriggerIDs.length
			    || pb.IsElemWithin(NeatUpload_LastEventSource, pb.TriggerIDs)))
		{
			pb.Display();
		}
		return true;
	});
						
	var eventsThatCouldTriggerPostBack = ['click', 'keypress', 'drop', 'mousedown', 'keydown'];
							
	for (var i = 0; i < eventsThatCouldTriggerPostBack.length; i++)
	{
		var eventName = eventsThatCouldTriggerPostBack[i];
		this.AddHandler(pb.FormElem, eventName, function (ev) {
			ev = ev || window.event;
			if (!ev)
			{
				return true;
			}
			var src = ev.srcElement || ev.target;
			if (!src)
			{
				return true;
			}
			NeatUpload_LastEventType = ev.type;
			NeatUpload_LastEventSource = src;
			NeatUpload_AlertShown = false;
			if (ev.type != 'click' && ev.type != 'keypress')
			{
				return true;
			}
			if (pb.IsElemWithin(src, NeatUpload_TriggerIDs)
			      || NeatUpload_TriggerIDs.NeatUpload_length == 0)
			{
				return true;
			}
			var tagName = src.tagName;
			if (!tagName)
			{
				return true;
			}
			tagName = tagName.toLowerCase();
			if (tagName == 'input' || tagName == 'button')
			{
				var inputType = src.getAttribute('type');
				if (inputType) inputType = inputType.toLowerCase();
				if (!inputType || inputType == 'submit' || inputType == 'image')
				{
					pb.ClearFileInputs(pb.FormElem);
				}
			}
			return true;
		}, true);
	}

	this.AddSubmittingHandler(pb.FormElem, function () {
		if (!NeatUpload_LastEventSource)
		{
			return;
		}
		if (pb.IsElemWithin(NeatUpload_LastEventSource, NeatUpload_TriggerIDs))
		{
			return;
		}
		if (NeatUpload_TriggerIDs.NeatUpload_length)
		{
			pb.ClearFileInputs(pb.FormElem);
		}
	});

	for (var i = 0; i < triggerIDs.length; i++)
	{
		NeatUpload_TriggerIDs[triggerIDs[i]] = ++NeatUpload_TriggerIDs.NeatUpload_length;
		this.TriggerIDs[triggerIDs[i]] = ++this.TriggerIDs.length;
	}
}

NeatUploadPB.prototype.Bars = new Object();

NeatUploadPB.prototype.IsElemWithin = function(elem, assocArray)
{
	while (elem)
	{
		if (elem.id && assocArray[elem.id])
		{
			return true;
		}
		elem = elem.parentNode;
	}
};

NeatUploadPB.prototype.CanCancel = function()
{
	try
	{
		if (window.stop || window.document.execCommand)
			return true;
		else
			return false;
	}
	catch (ex)
	{
		return false;
	}
}

NeatUploadPB.prototype.CombineHandlers = function(origHandler, newHandler) 
{
	if (!origHandler || typeof(origHandler) == 'undefined')
		return newHandler;
	return function(e) 
	{ 
		if (origHandler(e) == false)
			return false;
		return newHandler(e); 
	};
};

NeatUploadPB.prototype.AddHandler = function(elem, eventName, handler, useCapture)
{
	if (typeof(useCapture) == 'undefined')
		useCapture = false;
	if (elem.addEventListener)
	{
		elem.addEventListener(eventName, handler, useCapture);
		this.OnUnloadHandlers.push(function () { elem.removeEventListener(eventName, handler, useCapture); });
	}
	else if (elem.attachEvent)
	{
		elem.attachEvent("on" + eventName, handler);
		this.OnUnloadHandlers.push(function () { elem.detachEvent("on" + eventName, handler); });
	}
	else
	{
		var origHandler = elem["on" + eventName];
		elem["on" + eventName] = this.CombineHandlers(elem["on" + eventName], handler);
		this.OnUnloadHandlers.push(function () { elem["on" + eventName] = origHandler; });
	}
};

NeatUploadPB.prototype.EvaluateAutoStartCondition = function()
{
	with (this)
	{
		return eval(AutoStartCondition);
	}
}

NeatUploadPB.prototype.IsFilesToUpload = function()
{
	var formElem = this.FormElem;
	while (formElem && formElem.tagName.toLowerCase() != "form")
	{
		formElem = formElem.parentNode;
	}
	if (!formElem) 
	{
		return false;
	}
	var inputElems = formElem.getElementsByTagName("input");
	var foundFileInput = false;
	var isFilesToUpload = false;
	for (i = 0; i < inputElems.length; i++)
	{
		var inputElem = inputElems.item(i);
		if (inputElem && inputElem.type && inputElem.type.toLowerCase() == "file")
		{
			foundFileInput = true;
			if (inputElem.value && inputElem.value.length > 0)
			{
				isFilesToUpload = true;

				// If the browser really is IE on Windows, return false if the path is not absolute because
				// IE will not actually submit the form if any file value is not an absolute path.  If IE doesn't
				// submit the form, any progress bars we start will never finish.  
				if (navigator && navigator.userAgent)
				{
					var ua = navigator.userAgent.toLowerCase();
					var msiePosition = ua.indexOf('msie');
					if (msiePosition != -1 && typeof(ActiveXObject) != 'undefined' && ua.indexOf('mac') == -1
					    && ua.charAt(msiePosition + 5) < 7)
					{
						var re = new RegExp('^(\\\\\\\\[^\\\\]|([a-zA-Z]:)?\\\\).*');
						var match = re.exec(inputElem.value);
						if (match == null || match[0] == '')
						{
							if (typeof(NeatUpload_HandleIE6InvalidPath) != 'undefined'
							    && NeatUpload_HandleIE6InvalidPath != null)
								NeatUpload_HandleIE6InvalidPath(inputElem);
							return false;
						}
					}
				}
			}
		}
	}
	return isFilesToUpload; 
};

NeatUploadPB.prototype.ClearFileInputs = function(elem)
{
	var inputFiles = elem.getElementsByTagName('input');
	for (var i=0; i < inputFiles.length; i++ )
	{
		var inputFile = inputFiles.item(i);
		// NOTE: clearing (by removing and recreating) empty file inputs confuses IE6 when the document is
		// in both the top-level window and in an iframe.  ExpertTree uses such an iframe to do AJAX-style
		// callbacks.
		if (inputFile.type == 'file' && inputFile.value && inputFile.value.length > 0)
		{
			try
			{
				var newInputFile = document.createElement('input');
				for (var a=0; a < inputFile.attributes.length; a++)
				{
					var attr = inputFile.attributes.item(a); 
					if (attr.specified && attr.name != 'type' && attr.name != 'value')
					{
						if (attr.name == 'style' && newInputFile.style && newInputFile.style.cssText)
							newInputFile.style.cssText = attr.value;
						else
							newInputFile.setAttribute(attr.name, attr.value);
					}
				}
				newInputFile.setAttribute('type', 'file');
				inputFile.parentNode.replaceChild(newInputFile, inputFile);
			}
			catch (ex)
			{
				// I don't know of any other way to clear the file inputs, so on browser where we get an error
				// (eg Mac IE), we just give the user a warning.
				if (inputFile.value != null && inputFile.value != '')
				{
					if (!NeatUpload_AlertShown)
					{
						window.alert(this.ClearFileNamesAlert);
						NeatUpload_AlertShown = true;
					}
					return false;
				}
			}
		}
	}
	return true;
};

NeatUploadPB.prototype.AddSubmitHandler = function(elem, isPopup, handler)
{
	if (!elem.NeatUpload_OnSubmitHandlers) 
	{
		elem.NeatUpload_OnSubmitHandlers = new Array();
		elem.NeatUpload_OrigSubmit = elem.submit;
		elem.NeatUpload_OnSubmit = this.OnSubmit;
		try
		{
			elem.submit = function () {
				elem.NeatUpload_OnSubmitting();
				elem.NeatUpload_OrigSubmit();
				elem.NeatUpload_OnSubmit();
			};
			this.OnUnloadHandlers.push(function() 
			{
				elem.submit = elem.NeatUpload_OrigSubmit;
				elem.NeatUpload_OnSubmitHandlers = null;
				elem.NeatUpload_OnSubmit = null;
			});
		}
		catch (ex)
		{
			// We can't override the submit method.  That means NeatUpload won't work 
			// when the form is submitted programmatically.  This occurs in Mac IE.
		}			
	}
	if (isPopup)
	{
		elem.NeatUpload_OnSubmitHandlers.unshift(handler);
	}
	else
	{
		elem.NeatUpload_OnSubmitHandlers.push(handler);
	}	
};

NeatUploadPB.prototype.AddSubmittingHandler = function(elem, handler)
{
	if (!elem.NeatUpload_OnSubmittingHandlers) 
	{
		elem.NeatUpload_OnSubmittingHandlers = new Array();
		elem.NeatUpload_OnSubmitting = this.OnSubmitting;
		this.OnUnloadHandlers.push(function() 
		{
			elem.NeatUpload_OnSubmittingHandlers = null;
			elem.NeatUpload_OnSubmitting = null;
		});
	}
	elem.NeatUpload_OnSubmittingHandlers.push(handler);
};

NeatUploadPB.prototype.OnSubmitting = function()
{
	for (var i=0; i < this.NeatUpload_OnSubmittingHandlers.length; i++)
	{
		this.NeatUpload_OnSubmittingHandlers[i].call(this);
	}
	return true;
};

NeatUploadPB.prototype.OnSubmit = function()
{
	for (var i=0; i < this.NeatUpload_OnSubmitHandlers.length; i++)
	{
		if (!this.NeatUpload_OnSubmitHandlers[i].call(this))
		{
			return false;
		}
	}
	return true;
}	

NeatUploadPB.prototype.OnUnload = function()
{
	for (var i=0; i < this.OnUnloadHandlers.length; i++)
	{
		this.OnUnloadHandlers[i].call(this);
	}
	return true;
}	

NeatUpload_TriggerIDs = new Object();
NeatUpload_TriggerIDs.NeatUpload_length = 0;
NeatUpload_LastEventSource = null;
NeatUpload_LastEventType = null;
NeatUpload_AlertShown = true;

File: dotnet/src/Brettle.Web.NeatUpload/Brettle.Web.NeatUpload/ProgressBar.cs
/*

NeatUpload - an HttpModule and User Controls for uploading large files
Copyright (C) 2005  Dean Brettle

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

using System;
using System.Text;
using System.Collections;
using System.IO;
using System.ComponentModel;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.HtmlControls;
using System.Security.Permissions;
using System.Web.UI.Design;
using System.Drawing.Design;
using Brettle.Web.NeatUpload.Internal.UI;

namespace Brettle.Web.NeatUpload
{
	/// <summary>
	/// Displays progress and status of an upload.</summary>
	/// <remarks>
	/// For the progress bar to be displayed, the <see cref="UploadHttpModule"/> must be in use.
	/// For the progress display to be started, the form being submitted must include an <see cref="InputFile"/>
	/// control that is not empty.  Use the <see cref="Inline"/> property to control how the progress bar is
	/// displayed.  If you use the <see cref="Triggers"/> property (or the <see cref="AddTrigger"/> method)
	/// to specify which controls should cause files to be uploaded, then other controls will not start the progress
	/// display (e.g. "Cancel" buttons).
	/// </remarks>
	[AspNetHostingPermissionAttribute (SecurityAction.LinkDemand, Level = AspNetHostingPermissionLevel.Minimal)]
	[AspNetHostingPermissionAttribute (SecurityAction.InheritanceDemand, Level = AspNetHostingPermissionLevel.Minimal)]
	[DefaultProperty("Inline")]
	[ParseChildren(false)]
	[PersistChildren(true)]
	[ToolboxData("<{0}:ProgressBar runat='server'/>")]
	public class ProgressBar : ProgressBarBase
	{
		private bool InlineRequested
		{
			get { return (ViewState["inline"] != null && (bool)ViewState["inline"]); }
			set { ViewState["inline"] = value; }
		}
			
		/// <summary>
		/// Whether to display the progress bar inline or as a pop-up.  Under Opera, this property will always 
		/// return false, even if you set it to true.  Popup progress bars are automatically used under Opera 
		/// because Opera doesn't refresh the iframe used to display inline progress bars.</summary>
		[DefaultValue(false)]
		public bool Inline
		{
			get
			{
				// Return false when browser is Opera because Opera won't refresh the iframe until the upload completes.
				if (HttpContext.Current != null)
				{
					string userAgent = HttpContext.Current.Request.UserAgent;
					if (userAgent != null && userAgent.ToLower().IndexOf("opera") != -1)
						return false;
				}
				return InlineRequested;
			}
			set { InlineRequested = value; }
		}

        public bool AllowTransparency
        {
            get { return (ViewState["AllowTransparency"] != null && (bool)ViewState["AllowTransparency"]); }
            set { ViewState["AllowTransparency"] = value; }
        }

		private HtmlTextWriterTag Tag = HtmlTextWriterTag.Div;		
		
		protected override void OnPreRender (EventArgs e)
		{
			base.OnPreRender(e);

			if (!InlineRequested && !Width.IsEmpty && (Width.Type != UnitType.Pixel || Width.Value < 500))
				throw new System.ArgumentOutOfRangeException("Width must be at least 500 pixels and must use pixel(px) units when using a popup ProgressBar.");
			if (!InlineRequested && !Height.IsEmpty && (Height.Type != UnitType.Pixel || Height.Value < 100))
				throw new System.ArgumentOutOfRangeException("Height must be at least 100 pixels and must use pixel(px) units when using a popup ProgressBar.");

			if (Inline)
			{
				Tag = HtmlTextWriterTag.Iframe;
                Attributes["src"] = UploadProgressPath + LastPostBackIDQueryStringPortion + "&canScript=false&canCancel=false&postBackID=" + FormContext.Current.PostBackID;
				Attributes["frameborder"] = "0";
				Attributes["scrolling"] = "no";
				Attributes["name"] = this.ClientID;
                Attributes["allowTransparency"] = this.AllowTransparency.ToString();
			}
		}

		protected override string GetStartupScript()
		{
			string script = base.GetStartupScript();
			if (Inline)
				script += String.Format(@"<script type='text/javascript' language='javascript'>
<!--
NeatUploadPB.prototype.Bars['{0}'].DisplayUrl = function(progressUrl) {{
	var pb = this;
    var iframeWindow = document.getElementById(pb.ClientID).contentWindow || frames[pb.ClientID];
    iframeWindow.document.body.innerHTML = '';
	setTimeout(function () {{ iframeWindow.location.href = progressUrl; }}, 0);
}};

NeatUploadPB.prototype.Bars['{0}'].EvalOnClose = null;

(function() {{
	var pb = NeatUploadPB.prototype.Bars['{0}'];
    var iframeWindow = document.getElementById(pb.ClientID).contentWindow || frames[pb.ClientID];
	if (iframeWindow) 
		iframeWindow.location.replace(pb.UploadProgressPath + '{1}&postBackID=' + pb.UploadForm.GetPostBackID() + '&canScript=true&canCancel=' + NeatUploadPB.prototype.CanCancel());
}})();
// -->
</script>", ClientID, LastPostBackIDQueryStringPortion);
			return script;
		}

			
		public override void RenderBeginTag(HtmlTextWriter writer)
		{
			if (!Inline)
			{
				base.RenderBeginTag(writer);
				return;
			}
			
			if (!UploadModule.IsEnabled)
			{
				return;
			}
			base.AddAttributesToRender(writer);
			writer.RenderBeginTag(Tag);
		}
		
		protected override void RenderContents(HtmlTextWriter writer)
		{
			// Don't render the no-iframe fallback if the browser supports frames because there have been reports
			// that the fallback is briefly visible in some browsers.
			if (Inline && Context != null && Context.Request != null && Context.Request.Browser != null && Context.Request.Browser.Frames)
			{
		    	return;
		    }
			EnsureChildControls();
			base.RenderContents(writer);
		}
		
		public override void RenderEndTag(HtmlTextWriter writer)
		{
			if (!Inline)
			{
				base.RenderEndTag(writer);
				return;
			}
			if (!UploadModule.IsEnabled)
				return;
			
			writer.RenderEndTag();
		}

        private string LastPostBackIDQueryStringPortion
        {
            get
            {
                if (UploadModule.PostBackID == null)
                    return "";
                return "&lastPostBackID=" + UploadModule.PostBackID;
            }
        }
	}
}
